Mathlib/ProofTheory/LogicSymbol.lean

```diff
@@ -0,0 +1,431 @@
+/-
+Copyright (c) 2024 Shogo Saito. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shogo Saito. Adapted for mathlib by Hunter Monroe
+-/
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.FunLike.Basic
+/-!
+# Logic Symbols
+
+Proof Theory studies proofs as mathematical objects, including axiomatic (Hilbert), deductive,
+and sequent-based systems. This file defines logical symbols as objects abstractly to encompass
+these approaches. It also defines homomorphisms that preserve these connectives.
+
+## Main Definitions
+* `LogicalConnective` is defined so that `LogicalConnective F` is a type that has logical
+  connectives $\top, \bot, \land, \lor, \to, \lnot$.
+* `LogicalConnectiveHom` is defined so that `f : F →ˡᶜ G` is a homomorphism from `F` to `G` that
+preserves logical connectives.
+-/
+
+/-- Syntax typeclass for negation of formula -/
+@[notation_class] class FNot (α : Type*) where
+  /-- Negation of formula -/
+  fnot : α → α
+
+/-- Syntax typeclass for implication of formula -/
+@[notation_class] class FImp (α : Type*) where
+  /-- Implication of formula -/
+  fimp : α → α → α
+
+/-- Syntax typeclass for conjunction of formula -/
+@[notation_class] class FConj (α : Type*) where
+  /-- Conjunction of formula -/
+  fconj : α → α → α
+
+/-- Syntax typeclass for disjunction of formula -/
+@[notation_class] class FDisj (α : Type*) where
+  /-- Disjunction of formula -/
+  fdisj : α → α → α
+
+/-- Prefix notation for `fnot` -/
+prefix:75 "~" => FNot.fnot
+
+/-- Infix notation for `fimp` -/
+infixr:60 " ⭢ " => FImp.fimp
+
+/-- Infix notation for `fconj` -/
+infixr:69 " ⋏ " => FConj.fconj
+
+/-- Infix notation for `fdisj` -/
+infixr:68 " ⋎ " => FDisj.fdisj
+
+attribute [match_pattern]
+  FNot.fnot
+  FImp.fimp
+  FConj.fconj
+  FDisj.fdisj
+
+namespace ProofTheory
+
+/-- `LogicalConnective` class describes a type's logical connectives -/
+class LogicalConnective (α : Type*)
+  extends Top α, Bot α, FNot α, FImp α, FConj α, FDisj α
+
+/-- `DeMorgan` class describes proof systems implementing De Morgan's laws -/
+class DeMorgan (F : Type*) [LogicalConnective F] : Prop where
+  verum           : ~(⊤ : F) = ⊥
+  falsum          : ~(⊥ : F) = ⊤
+  imply (p q : F) : (p ⭢ q) = ~p ⋎ q
+  and (p q : F)   : ~(p ⋏ q) = ~p ⋎ ~q
+  or (p q : F)    : ~(p ⋎ q) = ~p ⋏ ~q
+  neg (p : F)     : ~~p = p
```

Comment:
What instances are you expecting of this class? Typical types of formulas will not satisfy these (unless you quotient by equivalence)